### PR TITLE
Upgrade to 1.1.1

### LINKED ISSUE #15: Data folder should be in yunohost.app
The data folder should be in /home/yunohost.app/funkwhale/data because sometimes the media can take lot of space fill up the limited space in /var. /home folder is generally reserved with lot of space to keep all the media. Further its good idea to keep media away from the app core files.
 I have faced this issue in Peertube too. 
See here https://github.com/YunoHost-Apps/peertube_ynh/issues/8

### PR DESCRIPTION
**Pull request by yalh76**
## Solution
- *Apply Example_ynh*
- *Fix #15*
- *Fix security issue*
- *Stick to official documentation installation*
- *Upgrade to 1.1.1*

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Package_check results
---
* An automatic package_check will be launch at https://ci-apps-dev.yunohost.org/, when you add a specific comment to your Pull Request: "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!"*

### CODE REVIEW

**Comment by yalh76** [comment]
!testme

File: conf/nginx.conf [hunk 1]
```diff
@@ -1,67 +1,145 @@
...
+  alias __DATADIR__/media/;
 }
 
-location /_protected/media/ {
-    # this is an internal location that is used to serve
-    # audio files once correct permission / authentication
-    # has been checked on API side
-    internal;
-    alias __FINALPATH__/media/;
+location /_protected/media {
```
Comment (Thovi98) [inline conf/nginx.conf:116]:
```suggestion
location /_protected/media/ {
```

File: conf/nginx.conf [hunk 2]
```diff
@@ -1,67 +1,145 @@
...
+# Comment the previous location and uncomment this one if you're storing
+# media files in a S3 bucket
+# location ~ /_protected/media/(.+) {
+#   internal;
+#   # Needed to ensure DSub auth isn't forwarded to S3/Minio, see #932
+#   proxy_set_header Authorization "";
+#   proxy_pass $1;
+# }
+
+location /_protected/music {
```
Comment (Thovi98) [inline conf/nginx.conf:133]:
```suggestion
location /_protected/music/ {
```

File: conf/nginx.conf (hunk 1 above)
Comment (Thovi98) [reply conf/nginx.conf:116]: Two suggestions to match, if I understand well, the CI.

Comment (yalh76) [reply conf/nginx.conf:116]: Not sure of that
Current configuration match the funkwhale nginx example.
It's a internal location that is used to serve audio files once correct permission / authentication has been checked on API side, so no risk of path transversal

It you have time to test with then ending `/` to be sure it's ok ...

**Comment by yalh76** [comment]
!testme

**Comment by Thovi98** [comment]
Oops sorry, I’ve just tested and media don’t play anymore. So I’ve removed the `/` and it works again, sorry for the trouble.

**Review by Thovi98 — APPROVED** [review]
Tested and approved

**Comment by Thovi98** [comment]
Mmmh, I’ve just updated my main instance, and I can’t listen to any audio anymore, behalve the podcasts. It seems that they are still in the database, but I’m not sure...
In the `funkwhale-server `logs, I’ve this : https://paste.yunohost.org/iqayazocey

**Comment by yalh76** [comment]
Can you look at nginx log errors ?

**Comment by Thovi98** [comment]
I have lots of 
```
2021/04/17 15:31:26 [error] 9306#9306: *1 open() "/var/www/funkwhale/front/dist/ynh_portal.js" failed (2: No such file or directory), client: 2a02:a03f:675c:6a00:88dd:c844:2abe:6b1a, server: funkwhale.domain.tld, request: "GET /ynh_portal.js HTTP/2.0", host: "funkwhale.domain.tld", referrer: "https://funkwhale.domain.tld/library/7914b51f-b4c2-4807-84a4-0bda216c8d92/upload?import=2021-04-17T15%3A29%3A25%2B02%3A00"
2021/04/17 15:31:26 [error] 9306#9306: *1 open() "/var/www/funkwhale/front/dist/ynh_overlay.css" failed (2: No such file or directory), client: 2a02:a03f:675c:6a00:88dd:c844:2abe:6b1a, server: funkwhale.domain.tld, request: "GET /ynh_overlay.css HTTP/2.0", host: "funkwhale.domain.tld", referrer: "https://funkwhale.domain.tld/library/7914b51f-b4c2-4807-84a4-0bda216c8d92/upload?import=2021-04-17T15%3A29%3A25%2B02%3A00"
2021/04/17 15:31:26 [error] 9306#9306: *1 open() "/var/www/funkwhale/front/dist/ynhtheme/custom_portal.js" failed (2: No such file or directory), client: 2a02:a03f:675c:6a00:88dd:c844:2abe:6b1a, server: funkwhale.domain.tld, request: "GET /ynhtheme/custom_portal.js HTTP/2.0", host: "funkwhale.domain.tld", referrer: "https://funkwhale.domain.tld/library/7914b51f-b4c2-4807-84a4-0bda216c8d92/upload?import=2021-04-17T15%3A29%3A25%2B02%3A00"
2021/04/17 15:31:26 [error] 9306#9306: *1 open() "/var/www/funkwhale/front/dist/ynhtheme/custom_overlay.css" failed (2: No such file or directory), client: 2a02:a03f:675c:6a00:88dd:c844:2abe:6b1a, server: funkwhale.domain.tld, request: "GET /ynhtheme/custom_overlay.css HTTP/2.0", host: "funkwhale.domain.tld", referrer: "https://funkwhale.domain.tld/library/7914b51f-b4c2-4807-84a4-0bda216c8d92/upload?import=2021-04-17T15%3A29%3A25%2B02%3A00"
2021/04/17 15:31:28 [error] 9306#9306: *1 open() "/home/yunohost.app/funkwhale/data/music/var/www/funkwhale/import/0. Musiques/0AD/You are Victorious! (Victory).mp3" failed (2: No such file or directory), client: 2a02:a03f:675c:6a00:88dd:c844:2abe:6b1a, server: funkwhale.domain.tld, request: "GET /api/v1/listen/d11ef963-9b8d-42f9-87ee-f76fa56104b3/?upload=eacdd38c-d2b4-403a-910f-3463b5af58d6&token=eyJ1c2VyX2lkIjoxLCJ1c2VyX3NlY3JldCI6Ijg5MGNlOTIzLTJjMzQtNGVmMS04MzVkLTFkNWI3YmM2MmJiMyIsInNjb3BlcyI6WyJyZWFkOmxpYnJhcmllcyJdfQ:1lXl21:mKYl7Dh4vRXir9G9_1m7IT5lCL8 HTTP/2.0", upstream: "http://127.0.0.1:5000/api/v1/listen/d11ef963-9b8d-42f9-87ee-f76fa56104b3/?upload=eacdd38c-d2b4-403a-910f-3463b5af58d6&token=eyJ1c2VyX2lkIjoxLCJ1c2VyX3NlY3JldCI6Ijg5MGNlOTIzLTJjMzQtNGVmMS04MzVkLTFkNWI3YmM2MmJiMyIsInNjb3BlcyI6WyJyZWFkOmxpYnJhcmllcyJdfQ:1lXl21:mKYl7Dh4vRXir9G9_1m7IT5lCL8", host: "funkwhale.domain.tld", referrer: "https://funkwhale.domain.tld/library/7914b51f-b4c2-4807-84a4-0bda216c8d92/upload?import=2021-04-17T15%3A29%3A25%2B02%3A00"
```

Last line is the issue I think. I’ve opted for the in place import, and so I’ve reimported the track, it doesn’t change anything. But when I delete the track from Funkwhale and import it again with the in place import, it works great again.

**Comment by yalh76** [comment]
it seems it look for the files at :
`/home/yunohost.app/funkwhale/data/music/var/www/funkwhale/import/0. Musiques/0AD/You are Victorious! (Victory).mp3`
instead of 
`/home/yunohost.app/funkwhale/data/music/0. Musiques/0AD/You are Victorious! (Victory).mp3`

**Comment by Thovi98** [comment]
Yes, I don’t know how we can better handle the migration from `/var/www/funkwhale/` to `/home/yunohost.app/funkwhale/data/music/` ?

Now by trying I’ve completely broken the app, I will clean install on testing branch.

**Comment by yalh76** [comment]
I've just tested the upgrade from current version to testing, and audio files are available ...

https://paste.yunohost.org/raw/gojivivoqa

**Comment by Thovi98** [comment]
Okay fine ! I hope I’ve had a minority issue. 
Do you think it can have a link with the fact that I had a symlink between my `/media/storage` and `/var/www/funkwhale` ?

**Comment by yalh76** [comment]
> 
> 
> Okay fine ! I hope I’ve had a minority issue.
> Do you think it can have a link with the fact that I had a symlink between my `/media/storage` and `/var/www/funkwhale` ?

Maybe but I'm like a newbie on linux... So you should check at your log files ...

**Comment by yalh76** [comment]
!testme

**Comment by Thovi98** [comment]
Here is my upgrade logs : https://paste.yunohost.org/raw/asobeqozuv
I didn’t find anything relevant, but I’m not good at reading logs...



> I'm like a newbie on linux...

Really ? You are awesome at packaging apps, which isn’t possible without a good knowing of Linux, no ? 😄

**Comment by yalh76** [comment]
Can you give me the path of you `You are Victorious! (Victory).mp3` file ? and the rights on it ?

Thanks, but packaging doesn't require a good knowledge of Linux:
- just being able to read a few bash
- stick to [example_ynh](https://github.com/YunoHost/example_ynh)
- search how other packages do things and copy/paste code from them (a good start is to search a word or a command in https://github.com/YunoHost-Apps )
- use and abuse of the YunoHost helpers
- being able to read YunoHost logs (quite the same as reading bash) or find where debian logs are
- being able to read and understand the installation process provided by the upstream apps

I only know a few linux commands: chmod, chown, cp, shutdown, reboot, journalctl, systemctl, nano, tail, that's all... What helps me: I'm a Windows© System admin: concepts are the same, implementation is just different .

**Comment by Thovi98** [comment]
Here is the path : `/media/stockage3/'0. Musiques'/0AD/You are Victorious! (Victory).mp3`

Oh okay, that’s interesssant, thanks for the information !

**Comment by yalh76** [comment]
> 
> 
> Here is the path : `/media/stockage3/'0. Musiques'/0AD/You are Victorious! (Victory).mp3`

It should be like `/home/yunohost.app/funkwhale/data//home/yunohost.app/funkwhale/data/media/tracks/59/30/2b/01-pardon.mp3`

So it seems you personalized where your music is stored and you should adapt `.env` file according at https://github.com/YunoHost-Apps/funkwhale_ynh/blob/f61a9c9cf48f45f432b3f75b05d9097f8d612554/conf/env.prod#L139-L140

and what are the rights on that file ?

**Comment by yalh76** [comment]
#Own CI result for Revision: f61a9c9cf48f45f432b3f75b05d9097f8d612554 

image

**Comment by Thovi98** [comment]
> So it seems you personalized where your music is stored and you should adapt `.env` file according at

I don’t think so, because I did a symlink a long time ago, like it’s proposed on the Readme.md : `ln -s /media/stockage3 /var/www/funkwhale/import`. It worked nice through the different updates...

The rights are actually `drwxr-xrwx` and owner is `nextcloud:nextcloud`, but it was maybe `root` before, I made recently some reconfigurations...

**Comment by Thovi98** [comment]
I will do a new test migration on a test instance with symlinking to my `/media/stockage3`.  If it works, I will consider my issue as particular and independent.

**Comment by yalh76** [comment]
arf.... I didn't manage the import folder... I will do a PR to correct that

**Comment by Thovi98** [comment]
Okay just did the new test and I could reproduce. So I will now remove and test again with your PR if you want :)

**Comment by yalh76** [comment]
just testing it before doing the commit

**Comment by yalh76** [comment]
should be fixed now: previous created symbolic links are move from `/var/www/funkwhale/import` to `/home/yunohost.app/funkwhale/data/music` but in the database, audio files are referenced with their full path ... Only option is to "import music from the server" two times ... first time, it cleanup imported audio files, second time, it do the import ...

I find way to do [an automatic import](https://docs.funkwhale.audio/admin/importing-music.html?highlight=import): but it requires to know the library ID to import audio files to....

```
export LIBRARY_ID="<your_libary_id>"
python api/manage.py import_files $LIBRARY_ID "`/home/yunohost.app/funkwhale/data/music/" --recursive --noinput
```

**Comment by Thovi98** [comment]
> `` but in the database, audio files are referenced with their full path ...

Is there any way to replace all `/var/www/funkwhale/import` occurence in the database ?

**Comment by yalh76** [comment]
> 
> Is there any way to replace all `/var/www/funkwhale/import` occurence in the database ?

I think it's possible doing SQL command in the postgresql database

**Comment by Thovi98** [comment]
https://www.postgresqltutorial.com/postgresql-replace/
It effectively seems possible, but I don’t understand how to do unfortunately...

**Comment by yalh76** [comment]
Importing two times from /import also works ... try it

**Comment by yalh76** [comment]
#Own CI results for Revision: a18d1a701dceaf7e0a4c5cfb114907aad2233a4e 
image


Ready to merge ?

**Comment by Thovi98** [comment]
If you think it’s good then it is for me :D 
But I fear that some poeple will lose there favorites and playlist if they have to reupload there files... Maybe a post on the forum to warn of a potential risk is a good idea ?

**Comment by yalh76** [comment]
> 
> 
> If you think it’s good then it is for me :D
> But I fear that some poeple will lose there favorites and playlist if they have to reupload there files... Maybe a post on the forum to warn of a potential risk is a good idea ?

I let you do the post, I'm just randomly going though funkwhale to solve the security issue ....
image

**Comment by Thovi98** [comment]
Okay, probably tonight :)

**Comment by Thovi98** [comment]
Just posted : https://forum.yunohost.org/t/funkwhale-update-to-1-1-1-revamp-be-cautious/15505

Don’t hesitate to change some text !

File: README_fr.md [hunk 3]
```diff
@@ -13,37 +13,24 @@ Un serveur de musique moderne, convivial et gratuit sur YunoHost
...
 * [Démo officielle](https://demo.funkwhale.audio)
 
-**Nom d’utilisateur :** demo  **Mot de passe :** demo
+**Nom d’utilisateur :** demo  **Mot de passe :** demo
 
 ## Admin
 
 L'administrateur utilise le login que vous avez fourni lors de l'installation. Le mot de passe est le même que celui que vous utilisez pour YunoHost.
 
 L'interface d'administration est accessible à l'adresse : votre.domaine.fr/api/admin
```
Comment (Thovi98) [inline README_fr.md:33]:
```suggestion
Pour ajouter une collection de fichiers musicaux à une bibliothèque dans votre installation YunoHost de Funkwhale, créez un lien symbolique vers votre collection intitulée "music" dans `/home/yunohost.app/funkwhale/data/music`.
`foo@bar:~$sudo ln -s /your/music/collection /home/yunohost.app/funkwhale/data/music`
Les fichiers peuvent ensuite être ajoutés à votre bibliothèque à partir de l'onglet *Envoi* dans une bibliothèque musicale sous la rubrique **Importer de la musique de votre serveur**.
```

File: README.md [hunk 4]
```diff
@@ -28,21 +28,13 @@ Installation requires a dedicated domain or subdomain. Installing in a subpath i
...
-
-  * [x] install/remove/backup/remove/upgrade with x86_64 and ARM
-  * [x] import file with web interface, imports from YouTube, interraction with MusicBrainz
-  * [x] [Subsonic API](https://docs.funkwhale.audio/users/apps.html)
-  * [x] [Federation](https://docs.funkwhale.audio/federation.html) tests
-  * [x] [CLI Import](https://docs.funkwhale.audio/importing-music.html#from-music-directory-on-the-server) Remember to `source $final_path/venv/bin/activate`, then `source $final_path/load_env` before typinh using manage.py.
-
-* to be added:
-  * [ ] Store files in HOME, see https://github.com/YunoHost-Apps/funkwhale_ynh/issues/15
+To add a collection of music files to a library in your YunoHost installation of Funkwhale, create a symlink to your collection titled "music" in `/home/yunohost.app/funkwhale/data`
```
Comment (Thovi98) [inline README.md:33]:
```suggestion
To add a collection of music files to a library in your YunoHost installation of Funkwhale, create a symlink to your collection titled "music" in `/home/yunohost.app/funkwhale/data/music`
```

**Review by Thovi98 — COMMENTED** [review]
Last but not least 😇

**Pull request merged**